File: C++/ValidParentheses.cpp
/**
 * Name: Muhammad Abdullah Uppal
 * Username: Abdullah-Uppal
 * Approach: Use stack to push opening brackets and whenever you get closing bracket match it with the most recently pushed opening bracket.
 * */
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;  
        for(auto i:s)  
        {
            if(i=='(' or i=='{' or i=='[') st.push(i);  
            else  
            {
                if(st.empty() || 
                    (st.top()=='(' && i!=')') || 
                    (st.top()=='{' && i!='}') || 
                    (st.top()=='[' && i!=']')
                ) return false;
                st.pop();  
            }
        }
        return st.empty(); 
    }
};

File: Python/ValidParentheses.py
# Name: Muhammad Abdullah

# Username: muhammadabdullah660

# Approach: Iteratively check the brackets/pattern of brackets in given string with dictionary key/values

class Solution(object):
    def isValid(self, s):
        stack = []
        dict = {"]": "[", "}": "{", ")": "("}
        for char in s:
            if char in dict.values():
                stack.append(char)
            elif char in dict.keys():
                if stack == [] or dict[char] != stack.pop():
                    return False
            else:
                return False
        return stack == []
